src/state/opamStateConfig.ml

```diff
@@ -190,14 +190,109 @@ let opamroot ?root_dir () =
    OpamStd.Env.getopt "OPAMROOT" >>| OpamFilename.Dir.of_string)
   +! default.root_dir
 
-let load opamroot =
-  OpamFile.Config.read_opt (OpamPath.config opamroot)
+(* XXX Keep here or move to OpamFile.Config *)
+let more_recent_raw version =
+  match version with
+  | Some v ->
+    OpamVersion.compare v OpamFile.Config.root_version > 0
+  | None -> false
+
+let more_recent config =
+    more_recent_raw (OpamFile.Config.opam_root_version config)
+
+let is_readonly_opamroot_raw ?(lock_kind=`Lock_write) version =
+  let newer = more_recent_raw version in
+  let write = lock_kind = `Lock_write in
+  if newer && write then None else
+    Some (newer && not write)
+
+let is_readonly_opamroot_t ?lock_kind gt =
+  is_readonly_opamroot_raw ?lock_kind
+    (OpamFile.Config.opam_root_version gt.config)
+
+let is_readonly_opamroot ?lock_kind gt =
+  is_readonly_opamroot_t ?lock_kind gt <> Some false
+
+let load_if_possible_raw ?lock_kind version (read,read_wo_err) f =
+  match is_readonly_opamroot_raw ?lock_kind version with
+  | None ->
+    OpamConsole.error_and_exit `Configuration_error
+      "Write lock and root more recent, abort!"
```

Comment:
As for `OpamGlobalState.with_write_lock` and `OpamRepositoryState.with_write_lock`